### PR TITLE
New Community module: GeoStyler

### PR DESCRIPTION
**Pull request by hwbllmnn**
## Description

This introduces the GeoStyler community module. See the mailing list for details.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)

The following are required only for core and extension modules, while they are welcomed, but not required, for community modules:
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by jodygarnett** [comment]
Quick note that CLA has been received.

**Comment by aaime** [comment]
As another note, the way it is setup it will be possible only for those building from sources to try it out.
Adding an assembly descriptor, while by no means a requirement, would allow the "adventurous users" to try it out from nightly builds, e.g.:
https://build.geoserver.org/geoserver/master/community-latest/

It's perfectly legit not to have it in there though, and you might also have reasons not to get it in the in the first place.

**Comment by hwbllmnn** [comment]
I added both the rudimentary docs and the assembly, thanks for the hints! I think this is ready to go now.

File: src/community/geostyler/src/main/resources/org/geoserver/wms/web/data/geostyler-init.ftl [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            style: response[0],
+            data: response[1],
+            compact: true,
+            enableClassification: false,
+            onStyleChange: onChange,
+            showAmountColumn: false,
+            showDuplicatesColumn: false
+        });
+
+        GeoStyler.locale.de_DE.GsRule.nameFieldLabel = 'Peter'
```
Comment (marcjansen) [inline src/community/geostyler/src/main/resources/org/geoserver/wms/web/data/geostyler-init.ftl]: 🤔

Comment (hwbllmnn) [reply src/community/geostyler/src/main/resources/org/geoserver/wms/web/data/geostyler-init.ftl]: I changed these to `GeoStyler`. Unsure where they actually appear in the UI (haven't seen them yet) though.

**Comment by hwbllmnn** [comment]
I'm unsure why travis keeps failing, looks like it has network connection problems?

**Comment by aaime** [comment]
https://travis-ci.community/t/install-of-oraclejdk11-is-failing-11-0-3-12/3887/3

**Comment by aaime** [comment]
I've tried to give it a spin with limited success.
The profile in web/app depends on a the wrong module name (geostyler vs gs-geostyler), once I've solved that locally, I get the tab but it's not working.

If I create a new style and switch immediately to the geostyler tab, I get the following:

```
java.lang.NullPointerException
	at org.geoserver.wms.web.data.GeoStylerTabPanel.renderHeaderScript(GeoStylerTabPanel.java:156)
	at org.geoserver.wms.web.data.GeoStylerTabPanel.renderHead(GeoStylerTabPanel.java:99)
	at org.apache.wicket.Component.renderHead(Component.java:4545)
	at org.apache.wicket.Component.internalRenderHead(Component.java:2751)
	at org.apache.wicket.markup.renderStrategy.ChildFirstHeaderRenderStrategy$1.component(ChildFirstHeaderRenderStrategy.java:85)
```

That is not surprising, as at that point the style file has not been created yet. Maybe the GeoStyler tab should not be visible on new style creation, but only after "apply" is done, just like the other tabs.

If I use an existing style instead, I get the following:

```
28 giu 09:28:01 WARN [RequestCycleExtra] - Handling the following exception
org.apache.wicket.WicketRuntimeException: The component(s) below failed to render. Possible reasons could be that:
	1) you have added a component in code but forgot to reference it in the markup (thus the component will never be rendered),
	2) if your components were added in a parent container then make sure the markup for the child container includes them in <wicket:extend>.

1. [HelpLink [Component id = styleGroupHelp, page = org.geoserver.wms.web.data.StyleEditPage, path = styleForm:context:panel:styleGroupHelp, type = org.geoserver.web.wicket.HelpLink, isVisible = true, isVersioned = false]]

	at org.apache.wicket.Page.checkRendering(Page.java:666)
	at org.apache.wicket.Page.endComponentRender(Page.java:317)
	at org.apache.wicket.page.XmlPartialPageUpdate.writeComponent(XmlPartialPageUpdate.java:127)
	at org.apache.wicket.page.PartialPageUpdate.writeComponents(PartialPageUpdate.java:248)
	at org.apache.wicket.page.PartialPageUpdate.writeTo(PartialPageUpdate.java:161)
	at org.apache.wicket.ajax.AjaxRequestHandler.respond(AjaxRequestHandler.java:358)
	at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:895)
	at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:64)
	at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:97)
	at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:265)
	at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:222)
	at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:293)
	at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:261)
	at org.apache.wicket.protocol.http.WicketFilter.processRequest(WicketFilter.java:203)
	at org.apache.wicket.protocol.http.WicketServlet.doPost(WicketServlet.java:159)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
```

This may not match your experience. I'm running GeoServer from the IDE (Start.java) activating the ``geostyler`` profile.

Anyways, the module compiles and the assembly portion succeeds in creating a zip file, so that's enough to merge the module, and you can fix the rest later if you want to. Just let me know how you want to proceed.

**Comment by hwbllmnn** [comment]
I fixed the dependency name, thanks for spotting. As for the other bugs, those were to be expected. The plugin currently doesn't work if you don't have a layer or a style yet. It would be great if you'd merge this for now, I expect we'll fix those issues down the road. Thanks!

**Pull request merged**